packages/hash/api/src/model/util.ts

```diff
@@ -51,4 +59,150 @@ export const RESTRICTED_SHORTNAMES = [
+ */
+export const workspaceAccountId = nilUuid;
+
+export const worskspaceTypesNamespaceUri =
+  "https://example.com/@workspace/types";
+
+export const blockprotocolTypesNamespaceUri =
+  "https://blockprotocol.org/@blockprotocol/types";
+
+type SchemaKind =
+  | EntityType["kind"]
+  | PropertyType["kind"]
+  | DataType["kind"]
+  | LinkType["kind"];
+
+const schemaKindSlugs: Record<SchemaKind, string> = {
+  entityType: "entity-type",
+  dataType: "data-type",
+  propertyType: "property-type",
+  linkType: "link-type",
+};
+
+const slugifySchemaTitle = (title: string): string =>
+  slugify(title.toLowerCase());
+
+export const generateSchemaBaseUri = (params: {
+  namespaceUri: string;
+  kind: SchemaKind;
+  title: string;
+}) =>
+  `${params.namespaceUri}/${schemaKindSlugs[params.kind]}/${slugifySchemaTitle(
+    params.title,
+  )}`;
+
+export const generateSchemaVersionedUri = (params: {
+  namespaceUri: string;
+  kind: SchemaKind;
+  title: string;
+  version?: number;
+}) => `${generateSchemaBaseUri(params)}/v/${params.version ?? 1}`;
+
+const primitiveDataTypeTitles = [
+  "Text",
+  "Number",
+  "Boolean",
+  "Empty List",
+  "Object",
+  "Null",
+] as const;
+
+export type PrimitiveDataTypeTitle = typeof primitiveDataTypeTitles[number];
+
+export const primitiveDataTypeVersionedUris = primitiveDataTypeTitles.reduce(
+  (prev, title) => ({
+    ...prev,
+    [title]: generateSchemaVersionedUri({
+      namespaceUri: blockprotocolTypesNamespaceUri,
+      kind: "dataType",
+      title,
+      /** @todo: get latest version of primitive data tyeps incase they are udpated */
+      version: 1,
+    }),
+  }),
+  {},
+) as Record<PrimitiveDataTypeTitle, string>;
+
+/**
+ * Helper method for generating a property type schema for the Graph API.
+ */
+export const generateWorkspacePropertyTypeSchema = (params: {
+  title: string;
+  /** @todo: account for nested property types (once we have a use-case) */
+  possibleValues: {
+    primitiveDataType: PrimitiveDataTypeTitle;
+    array?: boolean;
+  }[];
+}): PropertyType => ({
+  $id: generateSchemaVersionedUri({
+    namespaceUri: worskspaceTypesNamespaceUri,
+    title: params.title,
+    kind: "propertyType",
+  }),
+  kind: "propertyType",
+  title: params.title,
+  oneOf: params.possibleValues.map(({ array, primitiveDataType }) =>
+    array
+      ? {
+          type: "array",
+          items: {
+            oneOf: [
+              {
+                $ref: primitiveDataTypeVersionedUris[primitiveDataType],
+              },
+            ],
+          },
+        }
+      : {
+          $ref: primitiveDataTypeVersionedUris[primitiveDataType],
+        },
+  ),
+});
+
+/**
+ * Helper method for generating an entity schema for the Graph API.
+ */
+export const generateWorkspaceEntityTypeSchema = (params: {
+  title: string;
+  properties: {
+    baseUri: string;
+    versionedUri: string;
+    required?: boolean;
+    array?: boolean;
```

Comment:
nit and non-blocking:
If we suspect this will stay in the backend, should we type this such that `{ array: false, minItems: 3 }` is invalid? 

I think we can encode valid state instead such that the type of array becomes something like `array?: { minItems?: number, maxItems?: number } | boolean`. This is a very minor thing, and maybe obsolete with future changes. Just flagging :)

Reply:
This is potentially a temporary method, but I like your suggestion so have added it in https://github.com/hashintel/hash/pull/919/commits/8d7f12f5c374977ff652be977018d9f8fb25890c - thanks :)